libpgcommon/shared_gserialized.c

```diff
@@ -0,0 +1,98 @@
+/**********************************************************************
+ *
+ * PostGIS - Spatial Types for PostgreSQL
+ * http://postgis.net
+ *
+ * PostGIS is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * PostGIS is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PostGIS.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ **********************************************************************
+ *
+ * Copyright 2020 Raúl Marín
+ *
+ **********************************************************************/
+
+#include "lwgeom_cache.h"
+
+SHARED_GSERIALIZED *
+shared_gserialized_new_nocache(Datum d)
+{
+	SHARED_GSERIALIZED *s = palloc(sizeof(SHARED_GSERIALIZED));
+	s->count = 0;
+	s->geom = (GSERIALIZED *)PG_DETOAST_DATUM(d);
+	return s;
+}
+
+SHARED_GSERIALIZED *
+shared_gserialized_new_cached(FunctionCallInfo fcinfo, Datum d)
+{
+	SHARED_GSERIALIZED *s = MemoryContextAlloc(PostgisCacheContext(fcinfo), sizeof(SHARED_GSERIALIZED));
+	MemoryContext old_context = MemoryContextSwitchTo(PostgisCacheContext(fcinfo));
+	s->geom = (GSERIALIZED *)PG_DETOAST_DATUM_COPY(d);
+	MemoryContextSwitchTo(old_context);
+	s->count = 1;
+	return s;
+}
+
+SHARED_GSERIALIZED *
+shared_gserialized_ref(FunctionCallInfo fcinfo, SHARED_GSERIALIZED *ref)
+{
+	if (MemoryContextContains(PostgisCacheContext(fcinfo), ref))
```

Comment:
Do we have any take on the cost of `MemoryContextContains()`?

Reply:
Should be constant value. I (when doing this PR) learnt that pointers returned by palloc contain a header that precedes them, so all this does is check the header.

Reply:
👍